### DISCUSSION TITLE
Pick the display width as a limit when using the "custom size" option

### DISCUSSION THREAD

**Discussion by lulor** [rxhanson/Rectangle#491] [General]
(I initially opened this as [an issue](https://github.com/rxhanson/Rectangle/issues/490#issue-919545763) but I actually think that it would be more appropriate to discuss it here)

macOS version: Big Sur 11.4
Rectangle version: v0.47

I'm trying the new option to "Add an extra centering command with custom size" and I really like it, but I think I have encountered an issue:
if I specify a dimension that exceeds the display size in height, the window height gets correctly set as the display size (that acts as a limit), but if I do the same thing with the width, the window actually gets resized beyond the display.

A situation where this can be an issue is when the display scaling (and so its size in terms of number of pixels) changes "dynamically", for instance when plugging and unplugging a laptops from an external display.

I don't know if this is by design and/or it's fixable, but (in case it is) I personally think that it would make sense to pick the display width as a size limit.

Thank you for Rectangle!

**Comment by rxhanson** [comment]
Height is automatically limited by macOS when the height exceeds display height. This is probably so that you don't end up with windows behind the menu bar.

Are you looking to set a window a percentage of the display, rather than setting absolute size? If so, then you can modify Almost Maximize to be a different percentage size: https://github.com/rxhanson/Rectangle#almost-maximize.

Setting the absolute width could be desirable for some users - maybe someone would want a window that spans multiple displays?

I'll look to make the "extra centering command" accept values less than 1 and use those as percentages of the screen in a future release.

If you have any ideas that you are interested in implementing on this, it's pretty straightforward to dig in. The class where this calculation is performed is `SpecifiedCalculation`

**Reply by lulor** [reply]
Thank you for the explanation

Unfortunately I already use the Almost Maximize option for another custom size and I'm actually interested in setting a specific absolute window dimension..

Regarding having a single window spanning multiple displays: I didn't think about that use case, but does macOS allow you to do so? After a couple of tests it looks like a window can only be visible on a single display at a time, at least on my machine.

I'm not familiar with Swift, but I'll try having a look at the class you mentioned, thank you!

**Reply by rxhanson** [reply]
I went ahead and made the change to allow `SpecifiedCalculation` to accept values less than or equal to 1 for percentages. I typically release updates monthly at this point, though.

MacOS used to allow windows to span multiple displays (if I remember right), but you might be right that this is no longer possible. Perhaps it depends on the application. (I don't have my external display at the moment).

As a side note, my paid superset of Rectangle, [Hookshot](https://hookshot.app), is where I've built a custom shortcut creator for adding in custom sizes (Rectangle is purposefully simple, and the `SpecifiedCalculation` here is a community addition).

**Reply by lulor** [reply]
Thank you, I really appreciate it!

Looking at the code, I should be able to get the result I want to by just writing:

```
calculatedWindowRect.size.width = specifiedWidth <= 1  
        ? visibleFrameOfScreen.width * specifiedWidth
        : min( visibleFrameOfScreen.width, round(specifiedWidth) )
```

right?

**Reply by rxhanson** [reply]
Yeah, that should work

**Reply by lulor** [reply]
Thanks, I have just tried to build locally and it works as expected!

**Comment by lulor** [comment]
Hey @rxhanson, sorry for bothering you, but since apparently macOS doesn't allow you to have a window spanning multiple displays, would you consider making this the default behavior (just to know)?

**Reply by rxhanson** [reply]
I was actually going to include your change and I forgot (sorry!). I went ahead and just checked it in now so it'll be in the next release.

**Reply by lulor** [reply]
Thats's great to hear, thank you for your time!

**Reply by c-harding** [reply]
> since apparently macOS doesn't allow you to have a window spanning multiple displays

Have you tried turning off _Displays have separate Spaces_ in Mission Control preferences?

**Reply by raghavendra-talur** [reply]
Yes, when the displays have separate spaces option is turned off, you can have a window spanning multiple displays. @rxhanson would you be willing to add an behavior option for this?

**Reply by rxhanson** [reply]
It's on my todo list, but I don't have a guaranteed timeframe on when I will get to it. A good PR for this is welcome.
